cache/blobs_linux.go

```diff
@@ -0,0 +1,394 @@
+				if err := fs.Changes(ctx, filepath.Join(base, path), filepath.Join(upperdirView, path),
+					func(k fs.ChangeKind, p string, f os.FileInfo, err error) error {
+						return changeFn(k, filepath.Join(path, p), f, err) // rebase path to be based on the opaque dir
+					},
+				); err != nil {
+					return err
+				}
+				return filepath.SkipDir // We completed this directory. Do not walk files under this directory anymore.
+			}
+		}
+		return nil
+	})
+}
+
+// checkDelete checks if the specified file is a whiteout
+func checkDelete(upperdir string, path string, base string, f os.FileInfo) (delete, skip bool, _ error) {
+	if f.Mode()&os.ModeCharDevice != 0 {
+		if _, ok := f.Sys().(*syscall.Stat_t); ok {
+			maj, min, err := devices.DeviceInfo(f)
+			if err != nil {
+				return false, false, errors.Wrapf(err, "failed to get device info")
+			}
+			if maj == 0 && min == 0 {
+				// This file is a whiteout (char 0/0) that indicates this is deleted from the base
+				if _, err := os.Lstat(filepath.Join(base, path)); err != nil {
+					if !os.IsNotExist(err) {
+						return false, false, errors.Wrapf(err, "failed to lstat")
+					}
+					// This file doesn't exist even in the base dir.
+					// We don't need whiteout. Just skip this file.
+					return false, true, nil
+				}
+				return true, false, nil
+			}
+		}
+	}
+	return false, false, nil
+}
+
+// checkDelete checks if the specified file is an opaque directory
+func checkOpaque(upperdir string, path string, base string, f os.FileInfo) (isOpaque bool, _ error) {
+	if f.IsDir() {
+		for _, oKey := range []string{"trusted.overlay.opaque", "user.overlay.opaque"} {
+			opaque, err := sysx.LGetxattr(filepath.Join(upperdir, path), oKey)
+			if err != nil && err != unix.ENODATA {
+				return false, errors.Wrapf(err, "failed to retrieve %s attr", oKey)
+			} else if len(opaque) == 1 && opaque[0] == 'y' {
+				// This is an opaque whiteout directory.
+				if _, err := os.Lstat(filepath.Join(base, path)); err != nil {
+					if !os.IsNotExist(err) {
+						return false, errors.Wrapf(err, "failed to lstat")
+					}
+					// This file doesn't exist even in the base dir. We don't need treat this as an opaque.
+					return false, nil
+				}
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// sameDir performs continity-compatible comparison of directories.
+// https://github.com/containerd/continuity/blob/v0.1.0/fs/path.go#L91-L133
+// This doesn't compare files because it requires to compare their contents.
+// This is what we want to avoid by this overlayfs-specialized differ.
+func sameDir(f1, f2 os.FileInfo, f1fullPath, f2fullPath string) (bool, error) {
+	if !f1.IsDir() || !f2.IsDir() {
+		return false, nil
+	}
+
+	if os.SameFile(f1, f2) {
+		return true, nil
+	}
+
+	equalStat, err := compareSysStat(f1.Sys(), f2.Sys())
+	if err != nil || !equalStat {
+		return equalStat, err
+	}
+
+	if eq, err := compareCapabilities(f1fullPath, f2fullPath); err != nil || !eq {
```

Comment:
I don't think this is exactly what we think of equal xattrs. We should also keep everything in `user` scope. Can be follow-up.

Reply:
Let's address this in follow-ups.